Mathlib/Tactic/RewriteSearch.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2023 Kim Liesinger. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Liesinger
+-/
+import Mathlib.Data.List.EditDistance.Estimator
+import Mathlib.Data.MLList.BestFirst
+import Mathlib.Data.Nat.Interval
+
+/-!
+# The `rw_search` tactic
+
+`rw_search` attempts to solve an equality goal
+by repeatedly rewriting using lemmas from the library.
+
+If no solution is found,
+the best sequence of rewrites found before `maxHeartbeats` elapses is returned.
+
+The search is a best-first search, minimising the Levenshtein edit distance between
+the pretty-printed expressions on either side of the equality.
+(The strings are tokenized at spaces,
+separating delimiters `(`, `)`, `[`, `]`, and `,` into their own tokens.)
+
+The implementation avoids completely computing edit distances where possible,
+only computing lower bounds sufficient to decide which path to take in the search.
+
+# Future improvements
+
+We could call `simp` as an atomic step of rewriting.
+
+The edit distance heuristic could be replaced by something else.
+
+No effort has been made to choose the best tokenization scheme, and this should be investigated.
+Moreover, the Levenshtein distance function is customizable with different weights for each token,
+and it would be interesting to try optimizing these
+(or dynamically updating them,
+adding weight to tokens that persistently appear on one side of the equation but not the other.)
+
+Currently the `rw?` tactic (for generating a single rewrite) will not rewrite by a local hypothesis.
+Similarly, it will not use local hypotheses to discharge side conditions during rewriting.
```

Comment:
Is this still accurate once #6764 lands?

Reply:
```suggestion
The `rw_search` tactic will rewrite by local hypotheses,
but will not use local hypotheses to discharge side conditions. 
This limitation would need to be resolved in the `rw?` tactic first.
```

Reply:
Thanks! I will add a test for this.